### PR TITLE
[RFC] Migrate legacy test 3

### PR DESCRIPTION
**Pull request by glts**
This is a big one. The old test has about 50 'test cases' in almost 4000 lines (`.in` + `.ok` lines). Fortunately I was able to split these into `it` test cases in a semi-automated fashion. No changes to the test cases were necessary.

### CODE REVIEW

**Comment by Pyrohh** [comment]
May I ask how you automated the conversion (besides using the generic legacy2luatest)?

**Comment by glts** [comment]
I just used some Vim macros.

Cut the `test3.in`/`test3.ok` files into pieces (cut at `/^STARTTEST`, and at `/^\n\n` respectively), run `legacy2luatest.pl` on each of the files, `cat` all converted test files together. After this it was a just a matter of cleaning up: remove excess import statements, replace `describe` with `it`, sometimes add an empty first line to the test input. I did look through all tests though, just to be sure.

I know it's a lot to ask to review this. I think taking a good look at the general structure and taking a sample here and there should be good enough.

**Comment by Pyrohh** [comment]
I'll review this now, please don't rebase :)

**Comment by glts** [comment]
:smile_cat: I won't, thanks

**Comment by Pyrohh** [comment]
@glts sorry, but I've got to leave. I'm about half way through: due to it being so monotonous, it's not hard to review, but nevertheless very tedious :)

I made a bookmark and will finish it later today (It's only 16:22 currently).

**Comment by Pyrohh** [comment]
53 additional tests.... nice :smile:

Anyways, I'm going through it.

File: test/functional/legacy/003_cindent_spec.lua [hunk 1]
```diff
@@ -0,0 +1,4216 @@
...
+      
+      /* this is
+       * a real serious important big
+       * comment
+       */
+      	/* insert " about life, the universe, and the rest" after "serious" */
+      }
+      ]=])
+
+    execute('set tw=0 wm=60 columns=80 noai fo=croq')
```
Comment (Pyrohh) [inline test/functional/legacy/003_cindent_spec.lua]: is setting `columns` needed? From `:h columns`:
```
	Number of columns of the screen.  Normally this is set by the terminal
	initialization and does not have to be set by hand.  Also see
	|posix-screen-size|.
```
I was under the assumption that nvim had a virtual screen during these tests, that might be totally off though :)

File: test/functional/legacy/003_cindent_spec.lua [hunk 2]
```diff
@@ -0,0 +1,4216 @@
...
+      ]=])
+  end)
+
+  it('10 is working', function()
+    -- This is nasty. This is the only test case where the buffer content
+    -- differs from the original. Why? Proper behaviour of this test depends on
+    -- the fact that the setup code contains an (unbalanced) opening curly
+    -- bracket in "set cino={s,e-s". This bracket actually affects the outcome
+    -- of the test: without it the curly bracket under "void f()" would not be
+    -- indented properly. And that's why we've had to add one explicitly.
```
Comment (Pyrohh) [inline test/functional/legacy/003_cindent_spec.lua]: Thanks for commenting on this, as I had no clue what was going on :smile:

**Comment by Pyrohh** [comment]
Since this is one of the first tests I've reviewed, I just want to make sure I touched on everything I was supposed to:

- Verify the following:
  - temp files that were created were removed: `git grep -E 'os\.([a-z]*?)\(' test`
  - tests pass under same conditions as their equivalent legacy test (e.g. in terms of what is set via `execute()`)
  - tests were transferred correctly (icdiff was very helpful for this, but it's still mostly manual)
  - that the test count increases accordingly (I'm guessing that there's not much to worry about here given it's automated, idk though)
  - all `it()` blocks have a respective `expect()` block (again, not sure if there's not chance of an issue regarding this)
  - `before_each()` is used to avoid duplication
  - All tests pass :)

I really hope we don't have many more tests this big...

**Comment by tarruda** [comment]
> I just used some Vim macros.

Porting these tests are a great way to exercise macro skills :smile: 

Thanks for this great contribution @glts

File: test/functional/legacy/003_cindent_spec.lua (hunk 1 above)
Comment (glts) [inline test/functional/legacy/003_cindent_spec.lua]: Hm, I really have no idea.

I admit that I did not actually read the tests, I was happy with making it work.

**Comment by glts** [comment]
@Pyrohh Good list, and yes, I think that should about cover it. Thanks a lot for taking the time to review!

> > I just used some Vim macros.
> 
> Porting these tests are a great way to exercise macro skills :smile:

@tarruda :smile: yeah

And my method was probably quite inefficient, but with macros even the most stupid approach runs with rocket propulsion.

**Comment by glts** [comment]
Rebased.

File: test/functional/legacy/003_cindent_spec.lua (hunk 1 above)
Comment (glts) [inline test/functional/legacy/003_cindent_spec.lua]: I tried different `'columns'` settings and yes, it does affect the formatting (perhaps together with `'tw'` and `'wm'`?). Not sure it's worth the trouble investigating.

**Comment by justinmk** [comment]
restarted build

**Comment by glts** [comment]
hm, that's unfortunate, did it fail again even after restarting?

The failure doesn't seem to be anywhere near the legacy tests though.

**Comment by glts** [comment]
ok, I've just restarted it again, let's see

**Comment by glts** [comment]
Crap, the OS X build failed again but I swear it was the other one the last time, and had a lot more errors (just 1 this time).

Ok, I see the same is happening with other builds at the moment.

**Comment by justinmk** [comment]
Looks like `107_adjust_window_and_contents_spec.lua` has some indeterminism. Not related to this PR.

**Comment by glts** [comment]
Rebased with no changes.

The slowest build on Travis timed out. Just bad luck? I don't know, but this PR adds about 50 tests so if there is a fixed cost associated with each test then this might be a problem. If it is, then it's going to bite us sooner or later anyway.

**Comment by Pyrohh** [comment]
Do we actually know that Travis is timing out because of the tests?

**Comment by glts** [comment]
I think so, the log says that the timeout happened while running `functionaltest`.

**Comment by glts** [comment]
I think at this point we've established that none of the test failures were due to the changes in this pull request.

If somebody else wants to review this: you don't have to check all 4000 lines. As explained above, a sample here and there is probably enough.

**Comment by Pyrohh** [comment]
I'll take another look today, I can imagine you needing to rebase this repeatedly gets annoying :/

**Comment by glts** [comment]
No, no, I was just passing by ... this code is quite unfazed by what's going on elsewhere so no worries.

I should really go back to porting more tests though :|

**Comment by Pyrohh** [comment]
> I should really go back to porting more tests though :|

Why do we even pay you!

**Comment by glts** [comment]
:construction_worker: :sweat_drops:

**Comment by Pyrohh** [comment]
I went over it again using the checklist I mentioned above, and as far as I can tell it's exactly the same as last time I checked it. We should definitely wait for 1-2 more reviews given the size of htis... although I'm not sure anyone wants to review it :/

**Comment by Pyrohh** [comment]
Given that this is just a bunch of individual tests, perhaps they could be broken out into multiple smaller PRs?

**Comment by glts** [comment]
> Given that this is just a bunch of individual tests, perhaps they could be broken out into multiple smaller PRs?

Yeah, I suppose it could. I'm going to let this sit for now though, and perhaps revisit later. Thanks!

**Comment by justinmk** [comment]
Merged, thank you. I have no words... except thank you :) And @Pyrohh !

**Comment by Pyrohh** [comment]
```
 delete mode 100644 src/nvim/testdir/test3.in
 delete mode 100644 src/nvim/testdir/test3.ok
 delete mode 100644 src/nvim/testdir/test60.in
 delete mode 100644 src/nvim/testdir/test60.ok
 delete mode 100644 src/nvim/testdir/test60.vim
 delete mode 100644 src/nvim/testdir/test65.in
 delete mode 100644 src/nvim/testdir/test65.ok
 delete mode 100644 src/nvim/testdir/test76.in
 delete mode 100644 src/nvim/testdir/test76.ok
 delete mode 100644 src/nvim/testdir/test82.in
 delete mode 100644 src/nvim/testdir/test82.ok
 create mode 100644 test/functional/fixtures/autoload/footest.vim
 create mode 100644 test/functional/legacy/003_cindent_spec.lua
 create mode 100644 test/functional/legacy/060_exists_and_has_functions_spec.lua
 create mode 100644 test/functional/legacy/065_float_and_logic_operators_spec.lua
 create mode 100644 test/functional/legacy/076_completefunc_spec.lua
 create mode 100644 test/functional/legacy/082_string_comparison_spec.lua
```

... nice. Thanks a lot @glts and @justinmk, I was beginning to think this would end up like the fabled #2041 :grin:

**Comment by glts** [comment]
Thank you. One of these days we'll be able to say `git rm -rf src/nvim/testdir` or something ...

**Pull request closed without merging**